Project mng page

TakTypeR opened the pull request:
TypeAliasの件に関して、レビューをお願いします。
最新版は改変中の為、プロジェクトリストページの編集ボタンを押すとエラーが発生します。
よろしくお願いします。

TakTypeR committed:
feat: プロジェクトリスト管理用ページ追加。作成開始
doc: html id, Bootstrap, Thymeleafについて説明コメントを追加

TakTypeR committed:
feat: プロジェクト遷移管理追加
feat: プロジェクト管理ページ追加

TakTypeR committed:
Feat: プロジェクト一覧ページに項目内容を表示

TakTypeR committed:
Feat: 削除前確認の為、削除ダイアログをJQueryで作成(とりあえずVer)

TakTypeR committed:
Feat: CRUD実装の為、Project/ページにレコード削除機能を実装
Feat: 削除事故防止の為、Machine/削除前確認ダイアログを組み込み
Chore: ファイル名/project_view.js->common.js

TakTypeR committed:
Feat: Project/レコード編集機能作成

TakTypeR committed:
Feat: Project/新規レコード追加機能作成
Feat: サイドバー「プロジェクト登録」リンクを正式なページへ繋げた

TakTypeR committed:
Chore: Added database tables

TakTypeR committed:
Feat: Project/PC情報機能確認の為、関連PC情報保持機能を作成開始
Refactor: エンティティ記述簡易化の為、typeAlias機能をトライ

TakTypeR committed:
Fix: テスト側のプラットフォーム名を修正

TakTypeR committed:
Feat: プロジェクト-PC関係テーブルを使って、selectAllを再実装

TakTypeR committed:
Refactor: 未利用の為、機材情報のテーブル定義(pcs)からplatformを削除

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/entity/Project.java:
```diff
@@ -1,8 +1,44 @@
+    /**
+     * Constructors
+     */
+    public Project(){}
```
デフォルトコンストラクタは必要でしょうか？？
もし不要でしたら明記する必要は無いかと思いますので削除してしまっても問題ないかと！

TakTypeR replied:
習慣で記述していたのですが、現状は特に必要な状況もないので、削除します

YudaiNoda3576 replied:
デフォルトコンストラクタはJavaでは記述する必要がないので、明記しなくても問題ないかと思います！

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,49 @@
+     * @param id 検索対象の機材情報ID
+     * @return プロジェクト情報
+     */
+    Project findById(int id);
```
nullの可能性がある場合はOptionalでラップするのが安全かと思います！

TakTypeR replied:
そういう便利な仕組みがあるんですね。Optionalでラップする様にしました。
参考に質問です。
アプリ依存だと思いますが、これをContollerで呼び出してnullになった時(つまりID検索に失敗した場合)、どういう処理にするのが一般的ですか？(エラーダイアログを出して、ページ遷移させないとか？)

YudaiNoda3576 replied:
@TakTypeR 
そうですね、一般的かどうかは定かではなくて恐縮なのですが、
僕の触っていたプロジェクト（JavaはAPI、フロントはAngular）ではServiceクラスで例外のハンドリングをしてnullの場合は独自例外でエラーメッセージをフロントに返す構成にしてました。
レスポンスにエラーメッセージのJSONを乗せてフロントでエラーメッセージ’データがありません’的な物を表示させるようにしてました。

おそらく、404ページなどを表示させる場合はもっと深刻なエラーが発生した場合にして、値がある無い程度のエラーならばメッセージをレスポンスに返して表示させるくらいの方がユーザビリティは良いかもしれないです

答えになっていなくてすみません :pray:

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/service/ProjectService.java:
```diff
@@ -0,0 +1,62 @@
+public class ProjectService {
+
+    @Autowired
+    private ProjectMapper projectMapper;
```
コンストラクタインジェクションを個人的には推奨します！
参考：https://pppurple.hatenablog.com/entry/2016/12/29/233141

TakTypeR replied:
リンクありがとうございます。修正します
因みに、Serviceのみならず、Contoroller等クラスの役割に関わらず、基本的にコンストラクターインジェクションが推奨されるって事ですよね？

TakTypeR replied:
@RequiredArgsConstructorも良さそうなので、これも使ってみます

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/service/ProjectService.java:
```diff
@@ -0,0 +1,62 @@
+     * @param id 削除するプロジェクトID
+     * @return 削除後のDBのレコード数
+     */
+    public int deleteProject(int id)
```
更新系の処理は@Transactinalなどで明示的にトランザクションを行うのが良いかと！

YudaiNoda3576 replied:
また、更新処理が失敗した場合の処理なども検討するべきかなと思います

TakTypeR replied:
@Transactinal追加します。
Serviceクラスなので、メソッドでなくクラスに付与しようと思うのですが、正しいですか？

YudaiNoda3576 replied:
@TakTypeR 
基本的にはそれでも問題無いかと思います！
ただ、クラス単位で@Transactionalを付与した場合はクラス階層の上位の祖先クラスには適用されないなどの制約があるはずなので、使用する場合は注意が必要です。
僕はこの辺り考えるのが面倒臭いので、各メソッドに@Transactionaを付与するようにしています。
@Transactionalはpublicメソッドにしか適応されず、コンパイルエラーは起こらないので、そういった漏れを防ぐ意味でもメソッド単位で付与していくと少しは予防線になって良いかなと思って、、、

TakTypeR replied:
publicメソッドにのみ適用されるのは、覚えておきます
メソッド単位だと、親クラスや派生クラスの同名メソッドにも自動で適用されるという事でしょうか？

YudaiNoda3576 replied:
回答してませんでした！
> 親クラスや派生クラスの同名メソッドにも自動で適用される

これはどうなんでしょう...

YudaiNoda3576 commented on src/main/resources/application.properties:
```diff
@@ -14,7 +14,7 @@
 #spring.datasource.url=jdbc:mysql://localhost/userinfo?serverTimezone=JST
-spring.datasource.url=jdbc:mysql://localhost/project_management?serverTimezone=JST
+spring.datasource.url=jdbc:mysql://localhost/project_management?serverTimezone=JST&enabledTLSProtocols=TLSv1.2
 #���[�U��
```
秘密情報は環境変数などで隠蔽したいところですね。。。

TakTypeR replied:
url, username, passwordは環境変数を参照する様に変更しました。
コメントアウトしてある検証用のurlは、残してあります

YudaiNoda3576 commented on src/main/resources/templates/fragments/project/form_base.html:
```diff
@@ -0,0 +1,18 @@
+    </th:block>
+
+</body>
+</html>
```
改行をお忘れなくです！
IDEの設定で保存時に自動的に改行が入るように設定しておくのがおすすめです！

TakTypeR replied:
IDEの設定を変更しました

YudaiNoda3576 commented on src/main/resources/templates/machines.html:
```diff
@@ -84,12 +97,19 @@ <h6 class="m-0 font-weight-bold text-primary" th:text="|機材リスト(${numOfM
+    <script src="/js/common.js"></script>
+
 </body>
 </html>
```
こちらも改行ですね

TakTypeR replied:
同上

TakTypeR committed:
Refactor: インジェクションをコンストラクターインジェクションへ変更。
Feat: 検索系メソッドが失敗した時にnullを返すよう、Optionalへ変更
Feat: 更新系のメソッドがあるServiceクラスに@Transactionalを付与

TakTypeR committed:
Style: IDE上での文字化け解消の為、文字コードをUTF8へ変更
Feat: セキュリティの為、環境変数を参照する様にした

TakTypeR committed:
Style: 最終行に改行を追加。IDEの自動保存機能をON

TakTypeR committed:
Chore: Projectデバッグの為、プロパティファイルにログオプションを追加
Refactor: 不要なimportを削除

TakTypeR committed:
test: RDBから情報取得する部分デバッグ中

YudaiNoda3576 commented on src/main/resources/templates/projects.html:
```diff
@@ -0,0 +1,75 @@
+    <th:block th:insert="fragments/base::scripts"></th:block>
+    <script src="/js/common.js"></script>
+</body>
+</html>
```
こちらHTMLファイルにも改行をお忘れなくです〜！

TakTypeR replied:
挿入しました。
空白を入れる様に設定したのですが、そのファイルを変更等して、保存時に初めて適用されるんですかね？

YudaiNoda3576 commented on src/main/resources/mybatis-config.xml:
```diff
@@ -0,0 +1,12 @@
+        PUBLIC "-//mybatis.org//DTD Config 3.0//EN"
+        "http://mybatis.org/dtd/mybatis-3-config.dtd">
+<configuration>
+    <typeAliases>
```
typeAliasの問題は解決されたんでしたっけ？？
もし使用されていない場合は削除してしまうのが良いかと思います！

TakTypeR replied:
いえ。解決していないです。なので、削除しました

TakTypeR replied:
いえ。解決していないです。なので、設定を削除しました。

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/service/ProjectService.java:
```diff
@@ -0,0 +1,68 @@
+    /**
+     * 指定IDのプロジェクトを取得
+     */
+    public Optional<Project> getProject(int id) { return projectMapper.findById(id); }
```
Optionalで値が存在するか検証するメソッドが存在するのでその検証を挟む必要があります。
また、nullが取得された場合の異常系の処理も実装する必要がありますね！
 FYI
https://docs.oracle.com/javase/jp/8/docs/api/java/util/Optional.html#isPresent--

TakTypeR replied:
呼び出し側(editProject())では、Optional::ifPresentOrElse()で検証しています。
これで正しいんですよね？
異常系の処理は、別途行います。

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/service/ProjectService.java:
```diff
@@ -0,0 +1,68 @@
+ * プロジェクト情報に関するビジネスロジックを制御するクラス
+ */
+//更新系の処理を担当するクラスにはトランザクションを明示
+@Transactional
```
Servieceクラス全体をTransactionalに入れるのは個人的には非推奨です〜

TakTypeR replied:
了解です。
この場合、更新系の処理全般(deleteProject(), updateProject(), insertProject())に付与するのが正しいでしょうか？
ちなみに、非推奨の理由を教えて貰えますか？

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/service/ProjectService.java:
```diff
@@ -0,0 +1,68 @@
+    public int deleteProject(int id)
+    {
+        projectMapper.deleteById(id);
+        return numOfProject();
```
異常系の処理の実装も必要そうですね！
L.54も同様

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/controller/ProjectController.java:
```diff
@@ -0,0 +1,120 @@
+    @GetMapping("/project/edit/{id}")
+    public String editProject(@PathVariable("id") int id, Model model)
+    {
+        //Optional: nullになる可能性があるインスタンスはOptionalを使う
```
補足ですが、Optionalを使用することで、そのメソッドの実装者がnullが返却される可能性をコード上に明示できるというメリットがあります。（Optionalでラップされていれば、メソッドを使う人はnullが返る可能性があるんだなということが暗黙的にわかるので）
Optionalを使用しない場合、nullが返るかどうかをコメントで残していない限り、メソッドを呼び出す際に常にnullチェックを意識して実装を進めねばならず、常にヌルポが発生するリスクと隣り合わせで実装を進めねばならないということになります・

YudaiNoda3576 commented on src/main/java/com/example/raise_tech_lesson04/controller/ProjectController.java:
```diff
@@ -0,0 +1,120 @@
+        //Serviceクラスで例外のハンドリングをしてnullの場合は独自例外でエラーメッセージをフロントに返す構成にしていた。
+        //レスポンスにエラーメッセージのJSONを乗せてフロントでエラーメッセージ’データがありません’的な物を表示させるようにしてました。
+        //おそらく、404ページなどを表示させる場合はもっと深刻なエラーが発生した場合にして、
+        //値がある無い程度のエラーならばメッセージをレスポンスに返して表示させるくらいの方がユーザビリティは良いかもしれないです
```
ここにコメントされている形で実装のほどよろしくお願いします:pray:

YudaiNoda3576 reviewed (changes requested):
ザッとコメントしました！
次回からはPRの粒度を画面の実装単位（例えばエラーメッセージを部分的に表示する機能の実装）に分けるなどして頂くとより現場の運用に近づくかと思います！

TakTypeR committed:
Style: 最終行に空白行を追加
Refactor: 機能しないため、typeAliasesの設定を削除

TakTypeR committed:
Style: 最終行に空白行を追加

TakTypeR committed:
Fix:クラス全体をTransactionalに入れるのは非推奨の向きもあるため、メソッド単位へ変更(delete, update, insert)

TakTypeR committed:
Etc: DB接続には環境変数が優先される為、環境変数を使ってspring.datasource.urlを利用する方法に、コメントを追加。
Fixed errors on the unit test

TakTypeR committed:
Fix: MyBatisからProjectクラスへ、値が入らない不具合を修正
　　1.接続するDBがテスト用で、そちらはRDBが空だった
　　2.ProjectResultMap::machines::ColumnPrefixの使い方が間違っていた
Debug: テストコード追加

TakTypeR committed:
Feat: プロジェクトページの機能拡張。登録機材を同列に複数個表示する機能を作成。

TakTypeR committed:
Feat: プロジェクト所属の機材がある場合のみ、機材情報を出力する判定追加。ない場合は、メッセージを表示
Refactor: 記述をシンプルにする為、th:removeを使わず、spanを使って登録機材を同列に表示する方法へ変更
